### DISCUSSION TITLE
ref.listen don't listen to first state of state notifier provider

### DISCUSSION THREAD

**Discussion by thisisyusub** [rrousselGit/riverpod#1380] [Q&A] [answered]
I want to call some api when provider is initialized. This is the following:
<img width="581" alt="Screen Shot 2022-04-26 at 22 05 36" src="https://user-images.githubusercontent.com/44983544/165363957-e8fcf42e-cf0b-46fa-971e-731f91b28b12.png">
When I write `await Future.delayed(Duration.zero)` it works, listener working and dialog showing
<img width="703" alt="Screen Shot 2022-04-26 at 22 06 42" src="https://user-images.githubusercontent.com/44983544/165364126-5b5e8f55-8b00-474b-b662-76aff015c3fc.png">
UI code looks like this:
<img width="615" alt="Screen Shot 2022-04-26 at 22 07 52" src="https://user-images.githubusercontent.com/44983544/165364324-11e2abe0-7685-48b3-a0d8-de6b239ffdad.png">
But when delay is removed, listener is not working anywhere.

Otherhand, when I listen to `postFrameCallback` it is working as an expected.
<img width="662" alt="Screen Shot 2022-04-26 at 22 11 10" src="https://user-images.githubusercontent.com/44983544/165364869-a8a23acd-2b59-4f55-b8fe-5748b751a4bb.png">

How can I handle this situation in other way? I think these solutions are not good one.

**Comment by rrousselGit** [comment, marked as answer]
ref.listen triggers on change.

If you want to handle the very first value, read the initial value inside initState.

**Reply by thisisyusub** [reply]
So, using `postFrameCallback` is right solution?

**Reply by thisisyusub** [reply]
Btw, it is not inital value. It is the second state after initial state. I want to view loading in screen while api request is coming.

**Reply by rrousselGit** [reply]
Your snippet is incorrect. You shouldn't call `fetchRegistrationType`. You should check `isInProgress` and act accordingly.

> Btw, it is not inital value. It is the second state after initial state

It is the initial value. It's the value available at the time where `StateNotifierProvider`'s function returns.

**Reply by thisisyusub** [reply]
<img width="554" alt="Screen Shot 2022-04-26 at 22 58 58" src="https://user-images.githubusercontent.com/44983544/165372693-e810466b-f4a7-460c-8542-553d03bec1e6.png">

#### I think you mean this. But is it not overengineering? Because every time, in every page I should write this to handle first state. It is disappointed me. I have came from Bloc. Only initial state (super state) can not listen in bloc listener. is there no any solution like that?

**Reply by rrousselGit** [reply]
Your code is correct.
